### PR TITLE
feat: add terminal formatter

### PR DESCRIPTION
**Pull request by guilhermeprokisch**
Hi @Colonial-Dev,

Here is a draft for implementing the terminal formatter. My Rust skills are not the best, so there might be some bugs in it. I was able to make it work in one of my projects, so it should be at least on the right track. I'm sharing more to ensure that the work is not lost.

### CODE REVIEW

File: src/formatter/terminal_themed.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+///
+/// - The primary color of each style is used for the text color.
+/// - The secondary color is not used in this formatter.
+///
+/// This formatter holds theme data in an [`Arc<Theme>`], making it cheap to clone and usable
+/// from multiple threads.
+#[derive(Debug)]
+pub struct TerminalFormatter {
+    theme: Arc<Theme>,
+    stdout: RefCell<StandardStream>,
```
Comment (Colonial-Dev) [inline src/formatter/terminal_themed.rs]: `RefCell<T>` isn't `Send` or `Sync` (that is, multi-thread-compatible) like `Arc<T>` is.

For something like this, you *probably* don't want to be writing to standard output/error from multiple threads anyways, so I would just do something like:

```rs
#[derive(Debug, Clone)]
pub struct Terminal {
    theme: Theme,
    stream: StandardSteam
}
```

File: src/formatter/terminal_themed.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+/// A formatter for highlighting into terminal output with ANSI color codes.
+///
+/// - The primary color of each style is used for the text color.
+/// - The secondary color is not used in this formatter.
+///
+/// This formatter holds theme data in an [`Arc<Theme>`], making it cheap to clone and usable
+/// from multiple threads.
+#[derive(Debug)]
+pub struct TerminalFormatter {
```
Comment (Colonial-Dev) [inline src/formatter/terminal_themed.rs]: I would change this to just be `Terminal`, similar to how the basic HTML formatter is just called `Html`.

File: src/formatter/terminal_themed.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    theme: Arc<Theme>,
+    stdout: RefCell<StandardStream>,
+}
+
+impl TerminalFormatter {
+    /// Create a new [`TerminalFormatter`] with the given theme.
+    pub fn new(theme: Theme) -> Self {
+        Self {
+            theme: Arc::new(theme),
+            stdout: RefCell::new(StandardStream::stdout(ColorChoice::Always)),
```
Comment (Colonial-Dev) [inline src/formatter/terminal_themed.rs]: I wouldn't assume that users want standard output - consider something like printing code snippets in a backtrace to standard error. Maybe take a `StandardStream` as one of the parameters.

File: src/formatter/terminal_themed.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    fn write<W>(&self, source: &str, _writer: &mut W, event: HighlightEvent) -> Result<()>
+    where
+        W: std::fmt::Write,
+    {
+        match event {
+            HighlightEvent::Source { start, end } => {
+                let text = &source[start..end];
+                let mut stdout = self.stdout.borrow_mut();
+                stdout.write_all(text.as_bytes())?;
+                stdout.flush()?;
```
Comment (Colonial-Dev) [inline src/formatter/terminal_themed.rs]: I would implement `finish` and call `flush` in there - doing it for every event could get mighty expensive (unless the immediacy is important to your use case.)

**Review by Colonial-Dev — CHANGES_REQUESTED** [review]
Pretty good, thank you! I don't see any bugs, just a few things that need tweaking.

**Comment by guilhermeprokisch** [comment]
@Colonial-Dev, I have addressed some of your reviews. Please take a look and let me know if it meets your approval now. I have also added an example folder demonstrating how to use it, making it easy to see the results by running `cargo run --example terminal_highlight`. Additionally, I have included a theme generator. Frankly, this was the most challenging part, in my opinion. I believe that shipping with some built-in themes would be better.

**Review by Colonial-Dev — APPROVED** [review]

**Comment by Colonial-Dev** [comment]
Looks good to me!

> I believe that shipping with some built-in themes would be better.

Yeah, I feel you. The Helix theme support is almost done!

**Comment by guilhermeprokisch** [comment]
@Colonial-Dev  Btw I'm building a project based on this feature 

https://github.com/guilhermeprokisch/see/

Anything I'm here to help inkjet maintenance

**Pull request merged**